clash-prelude/src/Clash/Explicit/BlockRam/Model.hs

```diff
@@ -0,0 +1,311 @@
+  (Maybe a, Seq a)
+writeRam SNat isActiveWe updateMem addrX byteEnableX dat mem
+  -- Read (do nothing)
+  | IsDefined False <- isActiveWe byteEnableX
+  = (Nothing, mem)
+
+  -- Undefined address and write enable or (partially) unknown
+  | IsX addrMsg <- addrX
+  , let msg = "Unknown address" <> "\nAddress error message: " <> addrMsg
+  = ( Just (deepErrorX msg)
+    , Seq.fromFunction (natToNum @nAddrs) (unknownAddr msg) )
+
+  -- Write with defined address
+  | IsDefined addr <- addrX
+  = (Just dat, updateMem addr byteEnableX dat mem)
+ where
+  unknownAddr :: String -> Int -> a
+  unknownAddr msg n =
+    deepErrorX ("Write enabled or undefined, but address unknown; position " <> show n <>
+                "\nAddress error message: " <> msg)
+
+-- | Haskell model for a true dual port block RAM which is polymorphic in its
+-- write enables.
+--
+tdpbramModel ::
+  forall nAddrs domB domA a writeEnable .
+  ( HasCallStack
+  , KnownNat nAddrs
+  , KnownDomain domA
+  , KnownDomain domB
+  , NFDataX a
+  ) =>
+  -- | Determine whether a write enable is active
+  (MaybeX writeEnable -> MaybeX Bool) ->
+  -- | Merge global enable with write enable
+  (MaybeX Bool -> MaybeX writeEnable -> MaybeX writeEnable) ->
+  -- | Update memory with a defined address
+  (Int -> MaybeX writeEnable -> a -> Seq a -> Seq a) ->
+
+  Clock domA ->
+  -- | Enable
+  Signal domA Bool ->
+  -- | Address
+  Signal domA (Index nAddrs) ->
+  -- | Write enable
+  Signal domA writeEnable ->
+  -- | Write data
+  Signal domA a ->
+
+  Clock domB ->
+  -- | Enable
+  Signal domB Bool ->
+  -- | Address
+  Signal domB (Index nAddrs) ->
+  -- | Write byte enable
+  Signal domB writeEnable ->
+  -- | Write data
+  Signal domB a ->
+
+  (Signal domA a, Signal domB a)
+tdpbramModel
+  isActiveWe mergeWe updateMem
+  clkA enA addrA byteEnaA datA
+  clkB enB addrB byteEnaB datB =
+  ( startA :- outA
+  , startB :- outB )
+ where
+  (outA, outB) =
+    go
+      (Seq.fromFunction (natToNum @nAddrs) initElement)
+      (clockTicks clkA clkB)
+      (bundle (enA, byteEnaA, fromIntegral <$> addrA, datA))
+      (bundle (enB, byteEnaB, fromIntegral <$> addrB, datB))
+      startA startB
+
+  startA = deepErrorX $ "trueDualPortBlockRam: Port A: First value undefined"
+  startB = deepErrorX $ "trueDualPortBlockRam: Port B: First value undefined"
+
+  initElement :: Int -> a
+  initElement n =
+    deepErrorX ("Unknown initial element; position " <> show n)
+
+  go ::
+    Seq a ->
+    [ClockAB] ->
+    Signal domA (Bool, writeEnable, Int, a) ->
+    Signal domB (Bool, writeEnable, Int, a) ->
+    a -> a ->
+    (Signal domA a, Signal domB a)
+  go _ [] _ _ =
+    error "trueDualPortBlockRamModel#.go: `ticks` should have been an infinite list"
+  go ram0 (tick:ticks) as0 bs0 =
+    case tick of
+      ClockA -> goA
+      ClockB -> goB
+      ClockAB -> goBoth
+   where
+    (  toMaybeX -> enAx
+     , toMaybeX -> byteEnaAx0
+     , toMaybeX -> addrAx
+     , datA_
```

Comment:
The point is moot here, because it doesn't deshadow anything, but I've always started counting at `0`, not ∅. I figured that was what everyone was doing :sweat_smile:.

Reply:
Despite mention of a moot point being moot itself :-), I'll start with `0`s myself as well from now on. Although not in the name shadowing case...

```haskell
f x = g 42
 where
  g y = .... h ... x ...
  h x0 y z = ...
```

seems nicer then
```haskell
f x0 = g 42
 where
  g y = .... h ... x0 ...
  h x1 y z = ...
```
because these two `x`s arent part of some chain at all, the only reason the first arg of `h` needs a qualifier on its `x` is that there is already an `x` in scope. So changing the name of the binding made at the top level of this function to accomodate for an unrelated thing further on feels wrong to me. Clearly `x` is some actually descriptive name here, and not the `x`, `y`, `z` I'm using as placeholders. `x` could be `datA` and we'd nearly have the situation this discussion started with.